### PR TITLE
Traceviewer conversion script + Input Readiness metric

### PR DESCRIPTION
**Pull request by paulirish**
In this PR we have two items:

### Transform traceviewer source into consumable node module

* The PR builds on what @paullewis  did in #308
* Rationale: using the node_boostrap/vinn pipeline for traceviewer is a real hairball when it comes to browserifying. this simplifies a lot of concerns
* Updating to latest traceviewer source is manual. Added instructions to the readme.
* The transformed source is kept in `third_party/`. It spoils the diff of this PR, but it will be more maintainable in the long-run.

### Add the input readiness metric 

* Originally started in PR #284

Results of `lighthouse https://busrouter.sg/` (which is one of the busier PWAs):

image

Results of metric in the extension (via new browserified pipeline):

image

-------------------------------

_Tip to reviewers:_ Sorry about the large diff. The third_party files are at the end of the diff, so one you see them you can be done.

Also, this branch is on the origin repo. Commits appreciated, as it might be faster than the typical review cycle.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "third_party/src/catapult"]
```
Comment (brendankenny) [inline .gitmodules]: remove from `package.json` now?

File: scripts/convert.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Usage:
+ *   node scripts/convert.js
```
Comment (brendankenny) [inline scripts/convert.js]: rename to something more descriptive?

File: scripts/convert.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ */
+const fs = require('fs');
+const jsdom = require('jsdom');
+const mkdirp = require('mkdirp');
```
Comment (brendankenny) [inline scripts/convert.js]: need to add `jsdom` and `mkdirp` to deps

File: .gitmodules (hunk 1 above)
Comment (paulirish) [inline .gitmodules]: done

File: scripts/convert.js (hunk 2 above)
Comment (paulirish) [inline scripts/convert.js]: NO U

File: scripts/convert.js (hunk 3 above)
Comment (paulirish) [inline scripts/convert.js]: done

**Comment by paullewis** [comment]
I think we need to tell .eslint to ignore third_party, since we don't want to get into linting stuff there :)

**Comment by paulirish** [comment]
Input readiness is up and running now!

results of `lighthouse https://busrouter.sg/` which is one of the busier PWAs

image

**Comment by paulirish** [comment]
Currently this setup requires all LH developers initializing & updating a git submodule and running the convert script. 

An alternative is we keep third_party/traceviewer-js/ checked into the repo (not gitignored). We dont expect to bump it often and it would simplify things quite a bit. 

Thoughts?

**Comment by paullewis** [comment]
I'm fine with it. We update when new features land or something breaks. Go for it.

**Comment by paullewis** [comment]
Does it work in the extension?

**Comment by paulirish** [comment]
> Does it work in the extension?

ayup. 💃 

image

**Comment by paullewis** [comment]
Gosh dangit that report looks so retro nao.

**Comment by paulirish** [comment]
Ready for review

File: scripts/build-traceviewer-module.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    done: function(err, window) {
+      if (err) {
+        throw err;
+      }
+
+      const imports = window.document.querySelectorAll('link[rel="import"]');
+      const scripts = window.document.querySelectorAll('script');
+      let scriptsContent = '';
+
+      convertLicenseComments(html);
```
Comment (paullewis) [inline scripts/build-traceviewer-module.js]: We need some call to take across the LICENSE file from catapult and put it in the third_party folder, too.

Comment (paulirish) [inline scripts/build-traceviewer-module.js]: Cool. We can add that to the manual instructions in the bottom of the readme.

File: src/audits/performance/input-readiness-metric.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+const Audit = require('../audit');
+
+/* global window */
+window.global = window;
+
+// we need gl-matrix and jszip for traceviewer
+// since it has internal forks for isNode and they get mixed up during
+// browserify, we require them locally here and global-ize them.
+// from catapult/tracing/tracing/base/math.html
+var glMatrixModule = require('gl-matrix');
```
Comment (paullewis) [inline src/audits/performance/input-readiness-metric.js]: Can we shove all this into some other place and keep the funkiness out of the audit as much as possible?

Comment (paullewis) [inline src/audits/performance/input-readiness-metric.js]: We should also have some kind of 'unset' as well to remove these from the global.

**Comment by paullewis** [comment]
Reviewed.

File: scripts/build-traceviewer-module.js (hunk 4 above)
Comment (paullewis) [inline scripts/build-traceviewer-module.js]: Yeah it should be part of this PR though :)

File: src/audits/performance/input-readiness-metric.js (hunk 5 above)
Comment (paullewis) [inline src/audits/performance/input-readiness-metric.js]: nit: should be `const`

Comment (paulirish) [inline src/audits/performance/input-readiness-metric.js]: > Can we shove all this into some other place and keep the funkiness out of the audit as much as possible?

sure.  i moved it into tracing processor.

https://github.com/GoogleChrome/lighthouse/compare/paullewis-converter...paullewis-converter-refactorintotracingprocessor

want me to put this into this PR or do it as a followup?

File: scripts/build-traceviewer-module.js (hunk 4 above)
Comment (paulirish) [inline scripts/build-traceviewer-module.js]: done

**Comment by paulirish** [comment]
ptal

File: extension/gulpfile.js [hunk 6]
```diff
@@ -88,8 +88,11 @@ gulp.task('browserify', () => {
   ], {read: false})
     .pipe(tap(file => {
       file.contents = browserify(file.path, {
+        fullPaths: true,
```
Comment (brendankenny) [inline extension/gulpfile.js]: doesn't this really bulk up lighthouse-background.js? Especially with 200+ new files added to it

File: package.json [hunk 7]
```diff
@@ -34,20 +34,24 @@
...
     "npmlog": "^2.0.3",
     "semver": "^5.1.0",
-    "speedline": "^0.1.2",
-    "traceviewer": "^1.0.8"
+    "speedline": "^0.1.2"
   },
   "devDependencies": {
     "eslint": "^2.4.0",
     "eslint-config-google": "^0.4.0",
+    "gl-matrix": "2.3.1",
```
Comment (brendankenny) [inline package.json]: this is a dep, not a devDep?

File: scripts/build-traceviewer-module.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Usage:
+ *   node scripts/convert.js
```
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: `node scripts/build-traceviewer-module.js`

File: src/audits/performance/input-readiness-metric.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+require('../../../third_party/traceviewer-js/');
+const traceviewer = global.tr;
+
+class InputReadinessMetric extends Audit {
+  /**
+   * @override
+   */
+  static get tags() {
```
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: need to update to reflect #312. see e.g. [src/audits/html/meta-theme-color.js](https://github.com/GoogleChrome/lighthouse/blob/7fb2ecb922308fe5dd06af72b4a95bf5f6c6eac5/src/audits/html/meta-theme-color.js)

File: src/audits/performance/input-readiness-metric.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * @override
+   */
+  static get optimalValue() {
+    return '100';
+  }
+
+  /**
+   * Audits the page to give a score for First Meaningful Paint.
```
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: s/First Meaningful Paint/Input Readiness?

File: src/audits/performance/input-readiness-metric.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+      const events = [JSON.stringify({
+        traceEvents: artifacts.traceContents
+      })];
+      const model = new traceviewer.Model();
+      const importer = new traceviewer.importer.Import(model, io);
+      importer.importTraces(events);
+
+      // Now set up the user expectations model.
+      // TODO(paullewis) confirm these values are meaningful.
```
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: tracking issue for this?

File: src/audits/performance/input-readiness-metric.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+ */
+
+'use strict';
+
+const Audit = require('../audit');
+
+/* global window */
+window.global = window;
+
+// we need gl-matrix and jszip for traceviewer
```
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: can wait to integrate this with `tracing-processor.js` in a followup PR, but can you still pull this into a separate file (and require it here) so e.g. fixing fCP scoring isn't blocked while waiting for the followup changes?

File: scripts/build-traceviewer-module.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+      }
+
+      // adjust the javascript
+      for (let s = 0; s < scripts.length; s++) {
+        rewriteGlobals(scripts[s]);
+      }
+
+      writeNewFile(dest, scriptsContent);
+
+      function convertLicenseComments(html) {
```
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: passing in `html` but not `scriptsContent` seems weird here. Maybe just return the transformed `licenseContent` and append to `scriptsContent` at call site?

File: scripts/build-traceviewer-module.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+      }
+
+      function rewriteGlobals(script) {
+        script = script.textContent;
+
+        script = script.replace(/tr\.exportTo/, 'global.tr.exportTo');
+        script = script.replace(/var global = this;/, '');
+        script = script.replace(/this.tr =/, 'global.tr =');
+        script = script.replace(/\(function\(global\)\s?\{/, '(function() {');
+        scriptsContent += script;
```
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: return `script` and append external to function

File: src/audits/performance/input-readiness-metric.js [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+// from catapult/tracing/tracing/extras/importer/jszip.html
+global.JSZip = require('jszip/dist/jszip.min.js');
+
+global.HTMLImportsLoader = {};
+global.HTMLImportsLoader.hrefToAbsolutePath = function(path) {
+  if (path === '/gl-matrix-min.js') {
+    return 'empty-module';
+  }
+  if (path === '/jszip.min.js') {
+    return 'jszip/dist/jszip.min.js';
```
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: if you're requiring above, should this also return `empty-module`?

File: readme.md [hunk 16]
```diff
@@ -122,14 +122,22 @@ Promise.resolve({
...
 
 ## Trace processing
 
 The traceviewer-based trace processor from [node-big-rig](https://github.com/GoogleChrome/node-big-rig/tree/master/lib) was forked into Lighthouse. Additionally, the [DevTools' Timeline Model](https://github.com/paulirish/devtools-timeline-model) is available as well. There may be advantages for using one model over another.
+
+**To update traceviewer source:**
+
+```sh
+# if not already there, clone catapult and copy license over
+git clone --depth=1 https://github.com/catapult-project/catapult.git third_party/src/catapult
```
Comment (brendankenny) [inline readme.md]: can we pull this into a script in `scripts/`?

File: extension/gulpfile.js [hunk 17]
```diff
@@ -88,8 +88,11 @@ gulp.task('browserify', () => {
...
         transform: ['brfs']
       })
+      // explicitly require a few things traceviewer will need
+      .require(['gl-matrix', 'jszip/dist/jszip.min.js'])
```
Comment (brendankenny) [inline extension/gulpfile.js]: is this needed? why aren't these picked up from the `require`s in `input-readiness-metric.js`?

**Comment by brendankenny** [comment]
github really doesn't like your PR

**Comment by brendankenny** [comment]
`category` et al changes needed to fix the build

**Comment by brendankenny** [comment]
rebased and audit updated to fix the build

File: src/audits/performance/input-readiness-metric.js (hunk 9 above)
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: done

File: package.json (hunk 7 above)
Comment (brendankenny) [inline package.json]: done

File: scripts/build-traceviewer-module.js (hunk 8 above)
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: done

File: src/audits/performance/input-readiness-metric.js (hunk 10 above)
Comment (brendankenny) [inline src/audits/performance/input-readiness-metric.js]: done

File: scripts/build-traceviewer-module.js (hunk 13 above)
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: done

File: scripts/build-traceviewer-module.js (hunk 14 above)
Comment (brendankenny) [inline scripts/build-traceviewer-module.js]: done

File: src/audits/performance/input-readiness-metric.js (hunk 12 above)
Comment (paulirish) [inline src/audits/performance/input-readiness-metric.js]: This is done in https://github.com/GoogleChrome/lighthouse/commit/1984470ad17ac31e39e6734f7292d671362b10f8 which i'll land once this PR does.

File: extension/gulpfile.js (hunk 6 above)
Comment (brendankenny) [inline extension/gulpfile.js]: adds ~94KB to lighthouse-background.js, plus the local paths of whoever is building the extension. Running the extension, it appears to work without `fullPaths`, so, remove?

File: readme.md (hunk 16 above)
Comment (paulirish) [inline readme.md]: if it becomes a common thing yah

File: extension/gulpfile.js (hunk 6 above)
Comment (paulirish) [inline extension/gulpfile.js]: yah we dont need it.

Comment (brendankenny) [inline extension/gulpfile.js]: done

File: extension/gulpfile.js (hunk 17 above)
Comment (brendankenny) [inline extension/gulpfile.js]: done

**Comment by brendankenny** [comment]
LGTM with follow up PR

**Pull request merged**